pkg/backup/backupinfo/backup_index_test.go

```diff
@@ -0,0 +1,201 @@
+		{startTime: 8, endTime: 10},
+		{startTime: 4, endTime: 8},
+		{startTime: 2, endTime: 4},
+		{startTime: 0, endTime: 2},
+	}
+
+	var sortedFilenames []string
+	timeToFilenames := make(map[coverageTime]string)
+	for _, input := range inputs {
+		name := getBackupIndexFileName(
+			hlc.Timestamp{WallTime: input.startTime},
+			hlc.Timestamp{WallTime: input.endTime},
+		)
+		if _, exists := timeToFilenames[input]; exists {
+			t.Fatalf("duplicate index file name generated for %v", input)
+		}
+		timeToFilenames[input] = name
+		sortedFilenames = append(sortedFilenames, name)
+	}
+
+	slices.Sort(sortedFilenames)
+	expectedSortedFilenames := util.Map(
+		inputs,
+		func(input coverageTime) string {
+			return timeToFilenames[input]
+		},
+	)
+
+	require.Equal(
+		t, expectedSortedFilenames, sortedFilenames, "sort order of index filenames does not match",
+	)
+}
+
+func TestGetBackupIndexFilePath(t *testing.T) {
+	start, end := hlc.Timestamp{WallTime: 10}, hlc.Timestamp{WallTime: 20}
+	t.Run("fails if subdir is 'LATEST' and unresolved", func(t *testing.T) {
+		_, err := getBackupIndexFilePath("LATEST", start, end)
+		require.Error(t, err)
+	})
+	t.Run("returns correct path for resolved subdir", func(t *testing.T) {
+		subdir := "/2025/07/17-152115.00"
+		indexPath, err := getBackupIndexFilePath(subdir, start, end)
+		require.NoError(t, err)
+		require.True(
+			t, strings.HasPrefix(indexPath, path.Join(backupbase.BackupIndexDirectoryPath, subdir)),
+		)
+	})
+}
+
+func TestWriteBackupIndexMetadata(t *testing.T) {
+	ctx := context.Background()
+	externalStorage := newFakeExternalStorage()
+	makeExternalStorage := func(
+		_ context.Context, _ string, _ username.SQLUsername, _ ...cloud.ExternalStorageOption,
+	) (cloud.ExternalStorage, error) {
+		return externalStorage, nil
+	}
+
+	start := hlc.Timestamp{WallTime: 10}
+	end := hlc.Timestamp{WallTime: 20}
+	collectionURI := "nodelocal://1/backup"
+	subdir := "2025/07/18-143826.00"
+	backupPath := fmt.Sprintf("/incrementals/%s", subdir)
+
+	details := jobspb.BackupDetails{
+		Destination: jobspb.BackupDetails_Destination{
+			Subdir: subdir,
+		},
+		StartTime:     start,
+		EndTime:       end,
+		CollectionURI: collectionURI,
+		URI:           collectionURI + backupPath,
+	}
+
+	require.NoError(t, WriteBackupIndexMetadata(
+		ctx, username.RootUserName(), makeExternalStorage, details,
+	))
+
+	filepath, err := getBackupIndexFilePath(subdir, start, end)
+	require.NoError(t, err)
+
+	reader, nBytes, err := externalStorage.ReadFile(ctx, filepath, cloud.ReadOptions{})
+	require.NoError(t, err)
+
+	contents := make([]byte, nBytes)
+	_, err = reader.Read(ctx, contents)
+	require.NoError(t, err)
+
+	var metadata backuppb.BackupIndexMetadata
+	require.NoError(t, protoutil.Unmarshal(contents, &metadata))
+
+	require.Equal(t, start, metadata.StartTime)
+	require.Equal(t, end, metadata.EndTime)
+	require.Equal(t, backupPath, metadata.Path)
+}
+
+type fakeExternalStorage struct {
```

Comment:
I wanted some way of writing to "storage" and then reading it again so created this fake. Is there a better way of going about this? AI unfortunately was not able to point me to any existing code that might have something like this.